### PR TITLE
Document our code with doxygen

### PR DESCRIPTION
**Pull request by dkfellows**
This adds a build of doxygen documentation from our source code, and updates the documentation comments in our code to make reasonable use of this. See https://github.com/SpiNNakerManchester/sPyNNaker/pull/816 for sort of endpoint of where this PR is going.

Known Open Issues:
* Are we OK to auto-deploy the documentation? ([See here](http://spinnakermanchester.github.io/spinn_common/) for an approximate example of what it might look like; that's for `spinn_common` of course.) [**Answer:** Almost certainly. Benefit from doing so is great.]
* Should we also document the BMP? [**Answer:** YES! #126 addressed this]
   * What about as a separate build? (The BMP doesn't really share anything much with the rest of the tools after all.)

### CODE REVIEW

**Comment by dkfellows** [comment]
The single biggest change in all this is that I've converted a number of families of defines into enums (and updated the Perl code in `h2asm` to cope) so that things with a common group of meanings (e.g., command codes used in a particular field) are documented together.

File: include/spinnaker.h [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
-//
+//! \file  spinnaker.h
+//! \brief Header file describing SpiNNaker hardware
+//! \details See [SpiNNaker Datasheet
+//!     ](https://spinnakermanchester.github.io/docs/SpiNN2DataShtV202.pdf)
+//!
+//! \copyright &copy; The University of Manchester - 2009-2019
+//!
+//! \author Steve Temple, APT Group, School of Computer Science
+//! Email:            steven.temple@manchester.ac.uk
```
Comment (andrewgait) [inline include/spinnaker.h]: Does this email address for ST work any more?  (I note that in other places it has been removed...)

File: sark/sark_io.c [hunk 2]
```diff
@@ -409,17 +434,32 @@ void io_put_fixed(char *stream, uint n, uint d, uint a, uint pad, int neg)
 }
 
  // <drl add>
+//! \brief Format for s15.16 fixed point
```
Comment (andrewgait) [inline sark/sark_io.c]: Should this not be s16.15 ?

File: sark/sark_io.c [hunk 3]
```diff
@@ -409,17 +434,32 @@ void io_put_fixed(char *stream, uint n, uint d, uint a, uint pad, int neg)
...
+//! \param[in] n: signed accum, as signed integer
+//! \param[in] d: Field width
+//! \param[in] a: Precision
+//! \param[in] pad: Padding character
 static void io_put_sfixed(char *stream, int n, uint d, uint a, uint pad)
 {
     io_put_fixed(stream, (uint)((n < 0)? (-n): n) << 1, d, a, pad, (n < 0));
 }
 
+//! \brief Format for 16.16 fixed point
```
Comment (andrewgait) [inline sark/sark_io.c]: Is it worth clarifying that this is u16.16 ?

File: sark/sark_isr.c [hunk 4]
```diff
@@ -1,12 +1,22 @@
...
+//!
+//! \details
+//! In general, for the classes of interrupt with in this code, the options
+//! for handling are:
+//!
+//! 1. to dump the interrupt (clear the hardware but otherwise do nothing
+//! 2. to process the interrupt as a fast interrupt, using the FIQ (a very
+//!    limited number of event types can use this at a time)
+//! 3. to process the interrupt immediately as a standard interrupt.
+//! 4. to clear the hardware rapidly and enqueue a user-level callback for
```
Comment (andrewgait) [inline sark/sark_isr.c:17]: Nothing that needs changing.  Just like the word "enqueue", which I haven't seen used before IRL.

File: scamp/scamp-boot.c [hunk 5]
```diff
@@ -26,34 +25,40 @@
...
+};
 
-#define FF_START_PHASE_1                0x01000003
-#define FF_BLOCK_START_PHASE_1          0x02000003
-#define FF_BLOCK_DATA_PHASE_1           0x03000003
-#define FF_BLOCK_END_PHASE_1            0x04000003
-#define FF_CONTROL_PHASE_1              0x05000103
-
+//! Target monitor
+// TODO: What does this mean?
```
Comment (andrewgait) [inline scamp/scamp-boot.c]: Seems to be to do with nearest-neighbour targeting during boot, but if it's not on the datasheet then probably best leaving as a TODO.

**Review by andrewgait — CHANGES_REQUESTED** [review]
Much respect for reading the datasheet(s) and adding to the documentation.

Spotted a few minor things that need consideration.

File: include/spinnaker.h (hunk 1 above)
Comment (dkfellows) [reply include/spinnaker.h]: Forgot to remove that from the docs here.

File: sark/sark_io.c (hunk 2 above)
Comment (dkfellows) [reply sark/sark_io.c]: OK

File: sark/sark_io.c (hunk 3 above)
Comment (dkfellows) [reply sark/sark_io.c]: Done

File: scamp/scamp-boot.c (hunk 5 above)
Comment (dkfellows) [reply scamp/scamp-boot.c]: I _think_ it's about the same protocol as the standard flood fill. (There's an awful lot not stated on the datasheet.)

**Review by andrewgait — APPROVED** [review]
Resolutions of previous comments and the newly updated BMP doxygen code look fine to me.

**Comment by dkfellows** [comment]
Thank you for your help with reviewing.

**Pull request merged**